cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseConsumerStateStore.java

```diff
@@ -17,49 +17,122 @@
 
 import co.cask.cdap.api.common.Bytes;
 import co.cask.cdap.api.dataset.lib.AbstractDataset;
+import co.cask.cdap.api.dataset.table.Put;
 import co.cask.cdap.api.dataset.table.Row;
 import co.cask.cdap.api.dataset.table.Scanner;
 import co.cask.cdap.api.dataset.table.Table;
+import co.cask.cdap.common.collect.AllCollector;
+import co.cask.cdap.common.collect.Collector;
 import co.cask.cdap.common.queue.QueueName;
+import co.cask.cdap.data2.queue.ConsumerConfig;
+import co.cask.cdap.data2.queue.ConsumerGroupConfig;
 import co.cask.cdap.data2.transaction.queue.QueueConfigurer;
 import co.cask.cdap.data2.transaction.queue.QueueEntryRow;
+import co.cask.tephra.Transaction;
+import com.google.common.base.Charsets;
 import com.google.common.base.Preconditions;
-import com.google.common.base.Predicate;
+import com.google.common.collect.HashBasedTable;
 import com.google.common.collect.ImmutableMap;
-import com.google.common.collect.ImmutableSortedMap;
-import com.google.common.collect.Maps;
+import com.google.common.collect.Lists;
 import com.google.common.collect.Sets;
 import com.google.common.primitives.Ints;
 import com.google.common.primitives.Longs;
+import com.google.gson.Gson;
 
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Deque;
+import java.util.LinkedList;
+import java.util.List;
 import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+import javax.annotation.Nullable;
 
 /**
  * Class for persisting consumer state per queue consumer.
  */
 final class HBaseConsumerStateStore extends AbstractDataset implements QueueConfigurer {
 
+  private static final Gson GSON = new Gson();
+
   private final QueueName queueName;
   private final Table table;
-  private final byte[] minStartRow;
+  private final byte[] barrierScanStartRow;
+  private final byte[] barrierScanEndRow;
+  private Transaction transaction;
 
   HBaseConsumerStateStore(String datasetName, QueueName queueName, Table table) {
     super(datasetName, table);
     this.queueName = queueName;
     this.table = table;
-    this.minStartRow = QueueEntryRow.getQueueEntryRowKey(queueName, 0L, 0);
+    this.barrierScanStartRow = Bytes.add(queueName.toBytes(), QueueEntryRow.getQueueEntryRowKey(queueName, 0L, 0));
```

Comment:
the row key for the barrier is actually "queueName + queueEntryRow".. which is agnostic to what the actual format of the entry row key is (the actual entry row key is "flowletId + name + writeTx + counter").